src/coordinator.rs

```diff
@@ -168,41 +172,108 @@ impl Builder {
+        // dependent jobs before their dependencies. We can't do that because
+        // then we would have unresolved base keys and we'd have to defer until
+        // we got them or temporarily suffer incomplete information in the
+        // graph.
+        //
+        // Ideally, we'd look at the leaf nodes first, then the things that
+        // depend on them, etc. In other words, a depth-first search starting
+        // at the leaves instead of the roots. Lucky for us, that's easy to do:
+        // just write down the jobs we see as we do a depth first search, then
+        // walk that list in the opposite direction.
+        //
+        // `to_descend_into` tracks the depth-first search part of this scheme,
+        // and `to_convert` is where we write down the dependencies in root-to-
+        // leaf order.
+        let mut to_descend_into = self.roots.clone();
+        let mut to_convert = Vec::with_capacity(self.roots.len());
+
+        let mut glue_to_job_key: HashMap<&glue::Job, job::Key<job::Base>, Xxh3Builder> =
+            HashMap::with_capacity_and_hasher(self.roots.len(), Xxh3Builder::new());
+
+        let mut job_deps: HashMap<&glue::Job, HashSet<&glue::Job, Xxh3Builder>, Xxh3Builder> =
+            HashMap::with_hasher(Xxh3Builder::new());
+
+        while let Some(next_glue_job) = to_descend_into.pop() {
+            next_glue_job
+                .as_Job()
+                .inputs
+                .iter()
+                .filter(|item| item.discriminant() == glue::discriminant_U1::FromJob)
+                .for_each(|item| {
+                    let job = unsafe { item.as_FromJob() }.0;
+
+                    let entry = job_deps.entry(next_glue_job);
+                    entry
+                        .or_insert_with(|| HashSet::with_capacity_and_hasher(1, Xxh3Builder::new()))
+                        .insert(job);
+
+                    to_descend_into.push(job);
+                });
+
+            to_convert.push(next_glue_job);
+        }
+
+        while let Some(glue_job) = to_convert.pop() {
+            let job = job::Job::from_glue(glue_job, &glue_to_job_key)
+                .context("could not convert glue job into actual job")?;
+
+            if let Some(deps) = job_deps.get(glue_job) {
+                let blockers = coordinator.blocked.entry(job.base_key).or_default();
+
+                for dep in deps {
+                    blockers.insert(
+                        *glue_to_job_key
+                            .get(dep)
+                            .context("could not get job key for a glue job. This is probably an internal ordering bug and should be reported!")?
+                    );
+                }
+            } else {
+                coordinator.ready.push(job.base_key);
+            }
+
+            glue_to_job_key.insert(glue_job, job.base_key);
             coordinator.jobs.insert(job.base_key, job);
         }
 
+        for root in self.roots {
```

Comment:
I'm not sure I quite understand what this is doing. I guess it ends up mapping from the glue version of the roots to the regular version of the roots? And is needed due to the conversion from to_convert? I guess that makes sense. Just was unintuitive at first.

Reply:
yep, that's precisely it.

Reply:
I've added a comment explaining basically this